### PR TITLE
Fix treeo, treex, elegy

### LINKED ISSUE #159455: `python3Packages.treex` build failure on x86_64-linux as of `70aa8d9e`
## Issue description
Build of `python3Packages.treex` failed on x86_64-linux as of 70aa8d9ec6e80e9d6e5967c552ce665859863303.

[build log](https://github.com/samuela/nixpkgs-upkeep/actions/runs/1832991923)

Maintainer cc: @ndl
Other cc: n/a

### Steps to reproduce
1. Checkout nixpkgs at commit 70aa8d9ec6e80e9d6e5967c552ce665859863303
2. Run `nix-build -A python3Packages.treex`

## Technical details
```
- system: `"x86_64-linux"`
 - host os: `Linux 5.11.0-1028-azure, Ubuntu, 20.04.3 LTS (Focal Fossa)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.6.0`
 - nixpkgs: `/nix/store/f78jqkss0z39h2602xrrigzy6g1skpv6-source`
```

## Misc.
This issue was automatically generated by [nixpkgs-upkeep](https://github.com/samuela/nixpkgs-upkeep).
- [CI workflow](https://github.com/samuela/nixpkgs-upkeep/actions/runs/1832991923) that created this issue.
- Internal tag: 8597461721032410299

### PR DESCRIPTION
**Pull request by samuela**
###### Description of changes
Follow up to https://github.com/NixOS/nixpkgs/pull/159933. Fix https://github.com/NixOS/nixpkgs/issues/159455.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 166167` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.treeo</li>
    <li>python39Packages.elegy</li>
    <li>python39Packages.treeo</li>
    <li>python39Packages.treex</li>
  </ul>
</details>

**Comment by samuela** [comment]
I'll go ahead and merge tomorrow unless anyone objects.

File: pkgs/development/python-modules/treeo/default.nix [hunk 1]
```diff
@@ -8,21 +8,16 @@
 
 buildPythonPackage rec {
   pname = "treeo";
-  version = "0.4.0";
+  version = "0.0.9";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/treeo/default.nix:11]: We are intentionally downgrading?

Comment (samuela) [reply pkgs/development/python-modules/treeo/default.nix]: Upgrading treeo, end up involving pulling on a thread that required updating a few other packages. I just force-pushed a new sequence of commits with all of the new updates.

**Comment by samuela** [comment]
nixpkgs-review is claiming that I've updated 930 packages which is clearly false. But All of the packages that are being updated in this PR build successfully.

File: pkgs/development/python-modules/treeo/default.nix [hunk 2]
```diff
@@ -8,21 +8,18 @@
 
 buildPythonPackage rec {
   pname = "treeo";
-  version = "0.4.0";
+  # At the time of writing (2022-03-29) the latest version is 0.0.11 but that is
+  # incompatible with other packages, esp. treex. See https://github.com/cgarciae/treex/issues/68.
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/treeo/default.nix:12]:
```suggestion
  # 0.4.0 was an accidental release and was supposed to be 0.0.4
```
much more important

**Comment by SuperSandro2000** [comment]
If you move the rich update to another PR I think we can target master with the other things.

**Comment by samuela** [comment]
Is there a reason not to target rich to master?

**Comment by SuperSandro2000** [comment]
The 928 rebuilds on x86_64-linux.

**Comment by samuela** [comment]
> The 928 rebuilds on x86_64-linux.

Oh gotcha, fair enough!

File: pkgs/development/python-modules/treeo/default.nix (hunk 2 above)
Comment (samuela) [reply pkgs/development/python-modules/treeo/default.nix]: Fixed in https://github.com/NixOS/nixpkgs/pull/166167/commits/31da3adfc2a8ac2c302c2809486abbfc5895ad10.

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 166167` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>python310Packages.treeo</li>
    <li>python39Packages.augmax</li>
    <li>python39Packages.einops</li>
    <li>python39Packages.elegy</li>
    <li>python39Packages.flax</li>
    <li>python39Packages.optax</li>
    <li>python39Packages.treeo</li>
    <li>python39Packages.treex</li>
  </ul>
</details>

**Comment by samuela** [comment]
Ok I redid things, this time patching around the `rich` issue instead of updating it so as to keep this PR smaller and target master.

**Comment by samuela** [comment]
I also included the patch https://github.com/cgarciae/treeo/pull/14 to workaround issue https://github.com/cgarciae/treex/issues/68.

File: pkgs/development/python-modules/treeo/default.nix [hunk 3]
```diff
@@ -8,20 +9,26 @@
...
   };
 
-  postPatch = ''
-    substituteInPlace pyproject.toml \
-      --replace 'typing-extensions = "^3.10.0"' 'typing-extensions = "*"'
-  '';
+  # See https://github.com/cgarciae/treex/issues/68.
+  patches = [
+    (fetchpatch {
+      url = "https://github.com/cgarciae/treeo/pull/14.patch";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/treeo/default.nix:28]: Please fetch the commits individually to prevent errors when the PR is updated.

Comment (samuela) [reply pkgs/development/python-modules/treeo/default.nix]: Done in https://github.com/NixOS/nixpkgs/pull/166167/commits/db60f460b67463cdc12c3ad2982a49a80269717e

**Review by yl3dy — APPROVED** [review]

**Comment by samuela** [comment]
I'll go ahead and merge tomorrow unless anyone objects

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 166167` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
8 packages built:
</summary>
<ul>
<li>python310Packages.treeo</li>
<li>python39Packages.augmax</li>
<li>python39Packages.einops</li>
<li>python39Packages.elegy</li>
<li>python39Packages.flax</li>
<li>python39Packages.optax</li>
<li>python39Packages.treeo</li>
<li>python39Packages.treex</li>
</ul>
</details>

**Pull request merged**